afpacket/src/sockets.rs

```diff
@@ -0,0 +1,181 @@
+#![deny(missing_docs)]
+
+use libc;
+use std::{
+    io,
+    ptr,
+    mem::{self, MaybeUninit},
+    ffi::CStr};
+use crate::linux;
+
+/// Represents a link-local address.
+/// At this time, it's not particularly useful.
+pub struct Addr {
+    inner: libc::sockaddr_storage,
+    len: libc::socklen_t,
+}
+
+/// Represents an unbound `AF_PACKET` socket.  At this phase of a socket's lifecycle, it can be
+/// configured.
+pub struct Socket {
+    fd: libc::c_int,
+}
+
+/// Represents a bound `AF_PACKET` socket. At this phase of a socket's lifecycle, it can be read
+/// to/written from.
+pub struct BoundSocket {
+    fd: libc::c_int,
+    send_addr: libc::sockaddr_ll,
+}
+
+impl Socket {
+    /// Creates a new unbound socket.
+    pub fn new() -> io::Result<Self> {
+        // This block is unsafe because it invokes a syscall. It is safe because it checks the
```

Comment:
Is unsafe and is safe? Do you mean that the block below is unsafe but the `new` fn is safe?

Reply:
Sorry, this was a really poorly-worded comment. The idea I'm trying to communicate here is that while I'm using an `unsafe` block, the code in the `unsafe` block doesn't violate the borrow checker's invariants nor does it violate the type-system-enforced invariants of the `Socket` struct.

Reply:
Maybe we can communicate that with something like "The block below is mostly safe because A, B, C, but we use `unsafe` since we're making a `socket` syscall"?